_posts/2021-11-18-keras-updates/keras-updates.Rmd

```diff
@@ -0,0 +1,159 @@
+
+[^1]: Originally, both Keras and Tensorflow are Python libraries, bound to by R at runtime. In practice, this means that those Python libraries have to be present on the R user's machine. This is in contrast to R `torch` , which has no Python dependencies. (Instead, it directly binds to a C++ library.)
+
+So it is these three packages -- `tensorflow`, `tfdatasets`, and `keras` -- that should be understood by "Keras" in the current context[^2]. (The R-Keras ecosystem, on the other hand, is quite a bit bigger. But other packages, such as `tfruns` or `cloudml`, are more decoupled from the core.)
+
+[^2]: Definitely, `tfautograph` belongs in this group as well, providing even-lower-level support for already-low-level `tensorflow`. But normally, the user will not have to call its functions directly.
+
+Matching their tight integration, the aforementioned packages tend to follow a common release cycle, itself dependent on the underlying Python library, [TensorFlow](http://tensorflow.org). For each of `tensorflow`, `tfdatasets`, and `keras` , the current CRAN version is 2.7.0, reflecting the corresponding Python version. The synchrony of versioning between the two Kerases, R and Python, seems to indicate that their fates had developed in similar ways. Nothing could be less true, and knowing this can be helpful.
+
+In R, between present-from-the-outset packages `tensorflow` and `keras`, responsibilities have always been distributed the way they are now: `tensorflow` providing indispensable basics, but often, remaining completely transparent to the user; `keras` being the thing you use in your code. In fact, it is possible to train a Keras model without ever consciously using `tensorflow`.
+
+On the Python side, things have been undergoing significant changes, ones where, in some sense, the latter development has been inverting the first. In the beginning, TensorFlow and [Keras](keras.io) were separate libraries, with TensorFlow providing a backend -- one among several -- for Keras to make use of. At some point, Keras code got incorporated into the TensorFlow codebase. Finally (as of today), following an extended period of slight confusion, Keras got moved out again, and has started to -- again -- considerably grow in features.
+
+It is just that quick growth that has created, on the R side, the need for extensive low-level refactoring and enhancements. (Of course, the user-facing new functionality itself also had to be implemented!)
+
+Before we get to the promised highlights, a word on how we think about Keras.
```

Comment:
At this point I wish that you had at least mentioned what they were (in a word or two) somewhere in the introduction. I think it'd be worthwhile since, presumably, some people will want to read the post only to learn about the new features.

Reply:
Like I said in one other comment, I think i'd rather leave the narrative as-is ... We haven't written about Keras/TF for 2 years or so i guess, so this time, "Keras is back!" really is _the_ message.
Future posts will be more business-as-usual, again, then ...